core/src/main/java/org/infinispan/interceptors/impl/BatchingInterceptor.java

```diff
@@ -53,31 +50,36 @@ public Object visitEvictCommand(InvocationContext ctx, EvictCommand command) thr
     * suspend the batch's tx.</li> <li>If there is no batch in progress, just pass the call up the chain.</li> </ul>
     */
    @Override
-   protected Object handleDefault(InvocationContext ctx, VisitableCommand command) throws Throwable {
+   public CompletableFuture<Void> handleDefault(InvocationContext ctx, VisitableCommand command) throws Throwable {
       if (!ctx.isOriginLocal()) {
          // Nothing to do for remote calls
-         return invokeNextInterceptor(ctx, command);
+         return ctx.continueInvocation();
       }
 
       Transaction tx;
       if (transactionManager.getTransaction() != null || (tx = batchContainer.getBatchTransaction()) == null) {
          // The active transaction means we are in an auto-batch.
          // No batch means a read-only auto-batch.
          // Either way, we don't need to do anything
-         return invokeNextInterceptor(ctx, command);
+         return ctx.continueInvocation();
       }
 
       try {
          transactionManager.resume(tx);
-         InvocationContext invocationContext = ctx;
+         InvocationContext txInvocationContext = ctx;
          if (!ctx.isInTxScope()) {
-            // If there's no ongoing tx then BatchingInterceptor creates one and then invokes next interceptor,
+            // If there's no ongoing tx then BatchingInterceptor creates one and then invokes the chain again,
             // so that all interceptors in the stack will be executed in a transactional context.
-            // This is where a new context (TxInvocationContext) is created, as the existing context is not transactional: NonTxInvocationContext.
             log.tracef("Called with a non-tx invocation context: %s", ctx);
-            invocationContext = invocationContextFactory.createInvocationContext(true, -1);
+            txInvocationContext = invocationContextFactory.createInvocationContext(true, -1);
          }
-         return invokeNextInterceptor(invocationContext, command);
+         // Before sequential interceptors, we could continue the invocation with the next interceptor,
+         // with invokeNextInterceptor(txInvocationContext, command).
+         // But now we keep track of the invocation state (e.g. the current interceptor) in the invocation
+         // context itself (BaseSequentialInvocationContext, to be precise), so we have to restart the
+         // invocation with the new context instance.
+         // TODO Move the creation of the proper invocation context out of the interceptor and into CacheImpl
+         return ctx.shortCircuit(invoker.invoke(txInvocationContext, command));
```

Comment:
Yep. The assumption here, of course, is that the batching interceptor is the first one.